Scanner(unification) + TryPeek

AdamSpeight2008 opened the pull request:
This pull request unifies the contents of `TryScanToken` and `ScanTokenFullWidth`.
I retain those methods so not the break existing calls to them. The internals now call a common implementation method `ScanTokenCommon`.

This reduces the size of the sourcecode and the resulting IL (Debug)
```
                       Old |  New 
                     ------+------
      TryScanToken:-  1746 |   42
ScanTokenFullWidth:-  1972 |   15
   ScanTokenCommon:-       | 2035
                     ------+------
             Total:-  3718 | 2092
```

<!---
@huboard:{"order":3497.75,"milestone_order":1463,"custom_state":""}
-->

gafter commented:
@AlekseyTs Could you please look at this and decide if it is worth moving forward with it?

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
Resubmit within PR #1727 as that PR merge is simpler.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,37 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+    Friend Module CharExts
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsAnyOf(c As Char, c0 As Char, c1 As Char) As Boolean
```
Consider changing the name to something more descriptive. For example, IsEqualToAny.

AdamSpeight2008 replied:
What about if I add XML documentation to them instead?

AlekseyTs replied:
Unfortunately, XML documentation doesn't help at a call site unless you are reading the code in VS.

AdamSpeight2008 replied:
Considering that LNQ provides the extension method `.Any`, I don't the contextual meaning isn't that far of a leap.

AlekseyTs replied:
I am simply suggesting to name the methods in such a way so that no leap would be necessary.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,37 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        End Function
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsNoneOf(c As Char, c0 As Char, c1 As Char) As Boolean
```
Consider changing the name to something more descriptive. For example, IsNotEqualToAll.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/KeywordTable.vb:
```diff
@@ -304,9 +304,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
         Private Shared Function EnsureHalfWidth(s As String) As String
             Dim result As Char() = Nothing
-
+            Dim ch As Char
```
Is there a specific reason for this change? Consider reverting.

AdamSpeight2008 replied:
So it can be reused through out the loop.

AlekseyTs replied:
I don't believe moving the declaration out of the loop changes anything. What am I missing?

AdamSpeight2008 replied:
What is the cost of the declaration being redeclared on every iteration of the loop?

AlekseyTs replied:
None, I believe.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -222,13 +222,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                 ' // Don't break up surrogate pairs
-                Dim c = Peek()
-                Dim length = If(IsHighSurrogate(c) AndAlso CanGet(1) AndAlso IsLowSurrogate(Peek(1)), 2, 1)
+                Dim length = If(IsHighSurrogate(c) AndAlso TryPeek(1,c) AndAlso IsLowSurrogate(c), 2, 1)
```
Formatting: Consider adding a space after comma in TryPeek(1,c)

AdamSpeight2008 replied:
Done.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -471,10 +467,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             Debug.Assert(IsNewLine(StartCharacter))
 
-            Debug.Assert(StartCharacter = Peek(here))
+            Debug.Assert(StartCharacter = c)
```
It looks like new code is added for the purpose of the asserts. Consider surrounding the new variable declarations and the dependent asserts with #If DEBUG #End If

AdamSpeight2008 replied:
I've enclosed that entire section within a `#If Debug`

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -902,32 +883,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-            Debug.Assert(IsWhitespace(Peek()))
+            Dim c As Char
+            Dim ok = TryPeek(c)
+            Debug.Assert(ok)
```
Consider either inlining TryPeek(c) or making "Dim ok = TryPeek(c)" debug only code.

AdamSpeight2008 replied:
Made debug only.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -979,450 +960,191 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
+                Case CARRIAGE_RETURN, LINE_FEED
+                    Return ScanNewlineAsStatementTerminator(ch, precedingTrivia)
+                Case NEXT_LINE, LINE_SEPARATOR, PARAGRAPH_SEPARATOR
+                    If fullWidth Then Exit Select
```
Consider using multi-line If

AdamSpeight2008 commented:
@dotnet-bot test this please

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -979,450 +960,191 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                         ' TODO: do we allow widechars in keywords?
                         Dim spelling = "As"
-                        AdvanceChar(2)
+                        If fullWidth Then spelling = GetText(2) Else AdvanceChar(2)
```
Consider using multi-line If

AdamSpeight2008 replied:
I've had the feeling that this could be simplified to
```
Return MakeKeyword(SyntaxKind.AsKeyword, "If", precedingTrivia)
```
but never entirely sure.

AdamSpeight2008 replied:
@AlekseyTs I've tried both alternatives and I and I'm going with the first
```
AdvanceChar(2)
Return MakeKeyword(SyntaxKind.AsKeyword, "If", precedingTrivia)
```
As it calls less code.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -979,450 +960,191 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                         ' TODO: do we allow widechars in keywords?
                         Dim spelling = "End"
-                        AdvanceChar(3)
+                        If fullWidth Then spelling = GetText(3) Else AdvanceChar(3)
```
Consider using multi-line If

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -979,450 +960,191 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                         ' TODO: do we allow widechars in keywords?
                         Dim spelling = "If"
-                        AdvanceChar(2)
+                        If fullWidth Then spelling = GetText(2) Else AdvanceChar(2)
```
Consider using multi-line If

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1449,21 +1170,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-            Debug.Assert(CanGet)  ' > 
-            Debug.Assert(Peek() = ">"c OrElse Peek() = FULLWIDTH_GREATER_THAN_SIGN)
+            Dim c As Char
+            Dim ok = TryPeek(c)
```
Inline or #If DEBUG

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1476,26 +1197,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-            Debug.Assert(CanGet)  ' < 
-            Debug.Assert(Peek() = "<"c OrElse Peek() = FULLWIDTH_LESS_THAN_SIGN)
+            Dim c As Char
+            Dim ok = TryPeek(c)
```
Inline or #If DEBUG

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1571,13 +1290,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-            Debug.Assert(CanGet)
-            Debug.Assert(IsIdentifierStartCharacter(Peek))
+            Dim ch As Char
+            Dim ok = TryPeek(ch)
```
Inline or #If DEBUG

AlekseyTs replied:
Actually here we use result of TryPeek call in production code. Please ignore previous comment.

AdamSpeight2008 replied:
@AlekseyTs That's why some of them appear before the `Debug.Assert`'s so a second call is issued, where possible.

AdamSpeight2008 replied:
I've added in a comment here to explain that it has a Non DEBUG usage.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1689,34 +1403,35 @@ FullWidthRepeat:
-            Debug.Assert(CanGet)  ' [
-            Debug.Assert(Peek() = "["c OrElse Peek() = FULLWIDTH_LEFT_SQUARE_BRACKET)
+            Dim ch As Char
+            Dim ok = TryPeek(ch)
```
Inline or #If DEBUG

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -1776,10 +1493,9 @@ FullWidthRepeat:
+            If ch.IsAnyOf("&"c, FULLWIDTH_AMPERSAND) Then
                 Here += 1
-                ch = If(CanGet(Here), Peek(Here), ChrW(0))
+                If Not TryPeek(Here,ch) Then ch = ChrW(0)
```
Consider using multi-line If. Alternatively TryPeek could always assign ChrW(0) to the output parameter on failure, then 'if' wouldn't be needed.

AdamSpeight2008 replied:
I'm of the opinion that it should retain whatever it had on entry,

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -2154,31 +1841,17 @@ FullWidthRepeat2:
-               ByRef ReturnValue As Integer,
-               ByRef Here As Integer
-           ) As Boolean
+        Private Function ScanIntLiteral( ByRef ReturnValue As Integer, ByRef Here As Integer) As Boolean
```
Formatting: Consider removing space after '('

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -2508,44 +2177,35 @@ baddate:
 
             Dim length As Integer = 1
-            Dim ch As Char
+            Dim ch,cx As Char
```
Formatting: Consider adding a space after ','

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerBuffer.vb:
```diff
@@ -95,6 +95,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             Return p
         End Function
 
+        Friend Function TryPeek(at As Integer, ByRef ch As Char) As Boolean
```
Consider adding <Out> attribute to make the contract explicit.

AlekseyTs replied:
I mean Out attribute

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerBuffer.vb:
```diff
@@ -95,6 +95,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
+            ' CanGet(at)
+            Debug.Assert(_lineBufferOffset + at >= 0)
+            Debug.Assert(at >= -MaxCharsLookBehind)
+            If _lineBufferOffset + at >= _bufferLen Then Return False
```
Consider using multi-line If

AlekseyTs replied:
Also, ensure that ch is assigned to before the return.

AdamSpeight2008 replied:
@AlekseyTs Is not better to retain the value in `ch` and the False case?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerBuffer.vb:
```diff
@@ -95,6 +95,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
+            Return True
+        End Function
+
+        Friend Function TryPeek(ByRef ch As Char) As Boolean
```
Comments for the previous function is applicable fro this one too.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerInterpolatedString.vb:
```diff
@@ -112,22 +111,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-
+            Dim c, c1 As Char
+            If Not TryPeek(offset, c) Then Return False
+            Dim ok = TryPeek(offset + 1, c1)
```
Consider inlining TryPeek(offset + 1, c1), otherwise we might call it in cases when the call is not needed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerInterpolatedString.vb:
```diff
@@ -140,17 +138,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                 ' Any combination of fullwidth and ASCII curly braces of the same direction is an escaping sequence for the corresponding ASCII curly brace.
                 ' We insert that curly brace doubled and because this is the escaping sequence understood by String.Format, that will be replaced by a single brace.
                 ' This is deliberate design and it aligns with existing rules for double quote escaping in strings.
+                Dim ok = TryPeek(offset+1, c1)
```
Consider inlining TryPeek(offset + 1, c1), otherwise we might call it in cases when the call is not needed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerXml.vb:
```diff
@@ -786,7 +779,8 @@ CleanUp:
         Friend Function ScanXmlStringUnQuoted() As SyntaxToken
-            If Not CanGet() Then
+            Dim c As Char
+            If Not TryPeek(c) Then
```
It doesn't look like 'c' from this TryPeek call is used. Consider going back to CanGet.

AdamSpeight2008 replied:
Done and also move the declaration to be after the Assert.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerXml.vb:
```diff
@@ -1130,14 +1117,11 @@ CreateNCNameToken:
-            Debug.Assert(Peek() = "&"c)
-
+            Debug.Assert( NextIs(0,"&"c))
+            Dim c,c1 As Char
```
Formatting: Consider adding a space after the ','

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/XmlDocComments.vb:
```diff
@@ -506,12 +501,12 @@ CleanUp:
                     Case "<"c
-                        If CanGet(1) Then
-                            Dim ch As Char = Peek(1)
+                        Dim ch,c2 As Char
```
Formatting: Consider adding a space after ','

AlekseyTs commented:
Does this change also address #334?

AdamSpeight2008 commented:
@AlekseyTs It does address #334.

AdamSpeight2008 commented:
@AlekseyTs 
On the extension method names, let's put them to one side for a couple of days. To see any other alternative suggestions appear *(not say that your's are bad)*

AlekseyTs commented:
Sounds good.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerBuffer.vb:
```diff
@@ -95,6 +96,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             Return p
         End Function
 
+        Friend Function TryPeek(at As Integer, <Out()> ByRef ch As Char) As Boolean
```
Sorry, I didn't make it clear that two suggestions about this function were related.
If we want this function to follow TryGetValue pattern, which is a pretty well-known pattern, then 
1. Second parameter should be Out parameter (decorated with Out attribute)
2. Second parameter should be always assigned before the function returns.

Unless #2 is met, applying Out attribute is a violation of a contract for out parameters.

It is fine if we don’t want to follow the TryGetValue pattern, but then we shouldn’t have the Out attribute on the parameter. We should also explicitly state contract of the method in the comment. In addition we should consider slightly changing the name of the method so that consumers were not confused by the name similarity to TryGetValue. 

This comment also applies to the next function.

AdamSpeight2008 replied:
@AlekseyTs I'll have a look at this over the next few days, whilst I analyse reducing the number of additional variable usage. A job for my colour in pens, that way I'll see what affect of defaulting  to `ChrW(0)` on failure has.

AdamSpeight2008 replied:
Tried a version with `ChrW(0)` as the default but whilst still buildable it fails the unit tests.

AdamSpeight2008 commented:
@AlekseyTs I push a new commit to my branch, yet it hasn't updated the PR.

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
@gafter @AlekseyTs This builds and passes unit tests *(with the one regionalisation issue)* on my machine.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -902,32 +884,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-
+            Dim ch As Char
+            If Peep(i, ch) Then
+                ' Dim c3 As Char
```
Remove?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Scanner/XmlDocComments.vb:
```diff
@@ -282,22 +278,22 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-                        If CanGet(1) Then
-                            Dim ch As Char = Peek(1)
-                            Select Case ch
+                        '                    Dim c As Char
```
Remove?

AlekseyTs commented:
@dotnet-bot test this please

AlekseyTs commented:
LGTM. @gafter, @VSadov, @jaredpar, what do you think?

AdamSpeight2008 commented:
@AlekseyTs How do the names for renamed methods look?

AlekseyTs commented:
RE: How do the names for renamed methods look?
Look good to me.

VSadov commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,37 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsFrom(c As Char, c0 As Char, c1 As Char) As Boolean
+            Return (c = c0) OrElse (c = c1)
```
OrElse seems unnecessary here. OR with unconditional extra char compare is likely to be cheaper than deciding whether to do it or not.

AdamSpeight2008 replied:
@VSadov
The use of the `OrElse` adds only 2 IL bytes per a usage. So don't think that cost isn't too much.

VSadov commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -222,13 +222,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-
-            If Not CanGet() Then
+            Dim c As Char
+            If Not Peep(c) Then
```
CanGet used to be inlined by the JIT. Does Peep get inlined?

AdamSpeight2008 replied:
How would I check that?

VSadov replied:
I guess, we can assume that CanGet generally returns true and as such folding CanGet and following Peek would be  ok. 
In a case of "If Not CanGet()", CanGet seems would be substantially cheaper than "If Not Peep()"

VSadov replied:
The easiest is to try debugging Release with JIT optimizations not suppressed (setting under ToolsOptionsDebug) and take a look at x86/x64 at the calling site. 
Looking at the actual implementation of Peek it seems a bit too large to be an attractive inlining target.
As I commented above, For the cases where CanGet is a condition for the following Peek, folding that into a single Peep seems ok (since we assume that CanGet is mostly true). For the cases where the char is not used (like in negative tests), CanGet is likely to be substantially cheaper.

AdamSpeight2008 replied:
The IL size of the `Peep` methods are 91 and 96.

VSadov commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -979,450 +960,188 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                 Case "@"c
-                    Return MakeAtToken(precedingTrivia, False)
-
+                    Return MakeAtToken(precedingTrivia, fullWidth)
```
I think the goal for hardcoded "false" was the expectation that JIT would inline MakeAtToken and constant-propagate the "false", so only the codepath for "false" remains. Nonfullwidth scenarios are by far the most common case and the goal was to make it as fast as possible. Need to verify if that actually the case, since in such case the unified implementation could be slower.
Some scenarios in IDE can get fairly sensitive to the throughput of parsing/scanning, so any extra perf, even at the cost of maintainability (within reason) is considered a good tradeoff here, so must be careful that we are not doing the opposite.

AdamSpeight2008 replied:
It is only called by two methods (the pre-existing ones), call it with literals True and False.
[ScanTokenFullWidth](https://github.com/AdamSpeight2008/roslyn-AdamSpeight2008/blob/Scanner(Unification)/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L1135)
So wouldn't they still be "inlined"?
[TryScanToken](https://github.com/AdamSpeight2008/roslyn-AdamSpeight2008/blob/Scanner(Unification)/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L1143)

gafter replied:
@AdamSpeight2008 That couldn't hurt. I propose that we accept this PR (with that possible addition) and keep an eye on the performance for any regression.

gafter commented:
@AlekseyTs I propose that we accept this PR (with the possible addition of an attribute to encourage jit inlining) and keep an eye on the performance for any regression.

AdamSpeight2008 commented:
@gafter I'm just checking that option build and tests, once it passes I'll push a commit.

jaredpar commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        End Function
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsntFrom(c As Char, c0 As Char, c1 As Char) As Boolean
```
Prefer `IsNotFrom` over `IsntFrom`.  In general we don't use contractions in the APIs.

AdamSpeight2008 replied:
So you are saying that there is occasions that you use contractions. The APIs isn't public.
Using the contraction `Isnt` clearly separates it from `IsNot`

jaredpar commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        End Function
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsntFrom(c As Char, c0 As Char, c1 As Char, c2 As Char) As Boolean
```
Same

jaredpar commented on src/Compilers/VisualBasic/Portable/Scanner/CharacterInfo.vb:
```diff
@@ -488,4 +488,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        End Function
+
+        <Runtime.CompilerServices.Extension>
+        Friend Function IsntFrom(c As Char, c0 As Char, c1 As Char, c2 As Char, c3 As Char) As Boolean
```
Same

jaredpar commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -390,10 +386,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
 
 #Region "Buffer helpers"
 
+        <Runtime.CompilerServices.MethodImpl(Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)>
```
Why are we adding these attributes?  I don't think we should be using these unless we have a profile which proves their necessity.

AdamSpeight2008 replied:
I'll refer you to @gafter's reply.

> @AdamSpeight2008 That couldn't hurt. I propose that we accept this PR (with that possible > addition) and keep an eye on the performance for any regression.

jaredpar commented:
Why does this PR have some many uses of the inlining attribute?  Having those present without comment makes the code hard to decipher for a future developer.  My inclination in that scenario would be to simply delete them and move on.  

In general I don't think we should be adding them unless there is a profiler and a very strong reason to do so.

VSadov commented:
I agree with @jaredpar  on the forced inlining.
It makes sense to simplify common helpers to make them more attractive inlining targets, but we should not fight the JIT to inline methods.
It is very rare that we know something that JIT cannot know or there is a known flaw in the JIT's heuristics which might justify forced inlining.

gafter commented:
@AdamSpeight2008 The description of this PR is

> This pull request unifies the contents of TryScanToken and ScanTokenFullWidth.

But there are many other unrelated changes in the actual PR. Please remove all unrelated changes so we can review just that change.

gafter commented:
@AdamSpeight2008 The description of this PR is

> This pull request unifies the contents of TryScanToken and ScanTokenFullWidth.

But there are many other unrelated changes in the actual PR.

The scanner is the most performance sensitive area of the compiler. As such we are reluctant to change it and risk a performance regression. We are generally willing to put up with a suboptimal design (e.g. repeated code) in return for improved performance. Given where we are in the product cycle, we are reluctant to introduce instability in such a core component. Therefore we are skeptical of accepting this kind of change. Moreover, the fact that it has changed so much since being submitted (a dozen and a half revisions) feels like evidence that it isn't stable, and that volume of changes puts an unreasonable review burden on our team for very little customer benefit.

If we cannot converge soon on an acceptable form of this PR soon I would say we should put off these changes until after we are frozen for VS2015, and then at that time possibly reconsider the changes one-by-one.

AdamSpeight2008 commented:
@gafter  You told me to put the changes into this PR (see #1727)

jasonwilliams200OK commented:
@AdamSpeight2008 can you rebase this and squash the merge and fix commits?

AdamSpeight2008 commented:
@jasonwilliams200OK The merge looks really simple one conflict in `scanner.vb` that I think I've correctly fixed. Just waiting to see it the build passes the test. Don't know how to squash.

AdamSpeight2008 commented:
@jasonwilliams200OK done

jasonwilliams200OK commented:
@AdamSpeight2008, thanks. 

I think you still need to squash those commits:

(In case you don't know already)

Step 1:

```powershell
# powershell with git

cd projects\git\roslyn
git remote add Roslyn https://github.com/dotnet/roslyn
git pull --rebase roslyn master
git push -f
```

After this, step 2 will be needed for interactive rebasing (count may change, therefore, I will let you know the commands after step 1).

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -435,6 +437,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             Return result
         End Function
 
+
```
odd linefeed

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -578,9 +571,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-                If startComment = 0 AndAlso
-                    CanGet(Here) AndAlso
-                    Not IsNewLine(Peek(Here)) Then
+
```
odd linefeed

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerBuffer.vb:
```diff
@@ -95,6 +96,58 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
+            End If
+            Return True
+        End Function
+        ''' <summary>
```
require linefeed after `End Function`

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerXml.vb:
```diff
@@ -342,12 +341,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-                        If CanGet(1) Then
-                            Dim ch As Char = Peek(1)
+                        Dim ch As Char
+                        If Peep(1, ch) Then                            
```
whitespace after `Then`

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerXml.vb:
```diff
@@ -1033,9 +1021,9 @@ CleanUp:
-
-            Dim c = Peek(Here)
+            Dim c As Char
+            Dim ok =Peep(Here, c)
```
require whitespace before `Peep`

jasonwilliams200OK commented on src/Compilers/VisualBasic/Portable/Scanner/ScannerXml.vb:
```diff
@@ -1266,13 +1239,9 @@ CreateNCNameToken:
-                    Else
-                        Exit While
-                    End If
+
```
odd linefeed

AdamSpeight2008 commented:
@jasonwilliams200OK I think i messed up the rebase.

jasonwilliams200OK commented:
@AdamSpeight2008, what happened? can you retry https://github.com/dotnet/roslyn/pull/1463#issuecomment-107097616 (I had roslyn with capital R there, edited)

AdamSpeight2008 commented:
@jasonwilliams200OK 
Am I ready for step 2?

jasonwilliams200OK commented:
Step 2:

```powershell
# powershell with git

git rebase -i HEAD~15

#
# git interactive (-i) mode
#
# you have selected 15 commits on top of branch HEAD (which happen to be your own's commit).
# if you select more, you will probably messing with someone else's commits, so don't select
# more :)
#
# an editor such as notepad will open where you can organize the commits:
#
# - You can replace `pick` with `f` or `fixup` for every commit which you need to squash. This is
#   mostly you need to do. Note `fixup` is same as `squash` but it is better because it doesn't
#   produce the extra merge commit and yet produces a cleaner squash.
#
# - You can also rearrange the order of commit (but be careful, it will reapply the changes
#   in specified order and the final result may differ)
#
# - You can replace `pick` with `r` or `reword` for each commit to edit its log. Changing the
#   commit log there will not have effect. You will need to add `r`, save and close notepad
#   and then git will offer you to change log for each commit separately and open the
#   notepad/editor for each sequentially (close one to get the next one).
#
# - You can delete the commit line, this will remove that commit.
#
# for example:
#
# pick   Rename commonly used method to a shorter version. To aid in source      d234e4b  
# f      Rename CanGetChar() To CanGet()      6a76c61  
# pick   Implemented `AreNext( chars : string ) : Boolean` method overload.      844504b  
# pick   Added new extension methods, to for readability.      9d7c685  
# r      Some changes from comments on PR#1463      4b33c03  
# f      From PR#1463 comments      be54dec  
# f      Additional changes from PR comments     a16047f  
# r      Renamed the following methods      eac4463  
# f      Fix up XML Comments      13755a9  
# f      Remove the commented out code      7678021  
# pick   Add attribute to methods.      62c1863  
# f      Remove Attributes CharExts methods.      5d96898  
# f      Fixing bad merge Conflict in Scanner.vb      3a40ba7  
# f      Fix for Test Failure  a456c16  
# f      Fixing up the whitespace on the line ends, and where adding whitespace      0502b2a 
#
# Note that where i put `r` and `f`. I put `r` on those, which i do think you may not want to
# squash but keep, and the commit log should be improved. This is just an example.
#
# After that, save and close the notepad.
# If you have `r` or `reword` selected, for each reword, git will open the editor and let you
# change the message there.
#
# IMO, change all the commit messages (which you want to keep and not squash away)
# using this guideline: https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md#commit:
# (I wish someday dotnet team will inherit this guide someday, which is used as a standard in
# libuv,, node.js, npm and many major repositories! check their commit history to see the
# consistency in commit logs among thousands of commits..)
#
# - First line (commit heading) must be 50 chars long. The following lines (descript, which
#   appears on clicking the ellipsis `...` in commit history) must be 72 chars.
# - First part of first line (commit header) should be the feature name or the area
#   of affect, for example: say vb, or code, or compiler or even you can add multiple separated
#   by comma and followed by colon.
# - At the end of description, you can add Issue-URL, PR-URL and references etc.
#
# For example:
#
## --- commit message starts (this is not part of commit message itself) ---
#
# vb, compiler: Unifies scanner methods.
# The `TryScanToken` and `ScanTokenFullWidth` methods share most of
# the functionality. I have unified the code while keeping the signature
# to avoid backward compatibility.
#
# Issue-URL: #334
# Refs: #1727
# PR-URL: #1463
# Reviewed-By: @gafter, @AlekseyTs, @VSadov, @jaredpar
#
## --- commit message ends (this is not part of commit message itself) ---
#
# this is an example commit message for your first commit. And then you have
# other commits, for which you can use the same pattern.
#
# Once you save and close the last editor, and rebase is finished:
#

git push -f

# to overwrite/rewrite the history
```

And I think that's all there is to Step 2. -_-

AdamSpeight2008 commented:
@jasonwilliams200OK Done and Thank you for the notes on the rebase.

jasonwilliams200OK commented:
:+1: 
@AlekseyTs, all yours. IMO this PR (with highest number of comments) is ready to bring (probably little but) useful value to the roslyn code base. Please consider concluding it before it requires resolving merge conflicts again. :)

AdamSpeight2008 commented:
It makes the vb parser / scanner code a lot nicer to work with. Some section are little more efficient, some occur minor costs for the sake of clarity of code.

AlekseyTs commented:
@VSadov, @gafter, @agocke, @jaredpar, @khyperia FYI

AdamSpeight2008 commented on src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb:
```diff
@@ -995,450 +980,191 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
-                ch = MakeHalfWidth(ch)
-                Return ScanTokenFullWidth(precedingTrivia, ch)
+            If fullWidth Then
+                Debug.Assert(Not IsDoubleQuote(ch))
```
Debating whether it is worth making this as debug only? It be an empty branch (true) body in release mode..

gafter commented:
I think the concerns about overuse of AggressiveInlining should be addressed.

I agree with Jared that we should avoid contractions in method names.

I am concerned that this PR combines too many changes to be effectively reviewed.

jaredpar commented:
I agree with @gafter

AdamSpeight2008 commented:
@gafter Strange I thought I'd already addressed those, unless the squash zapped them.

AdamSpeight2008 commented:
@dotnet-bot test this please

AdamSpeight2008 commented:
@gafter @jaredpar @AlekseyTs @jasonwilliams200OK @ljw1004 

Hasn't this been slowly reviewed? over the course of its inception to now, through it varied incarnations. A little tweak there another tweak here, since it involve a delicate part of the VB.net compiler. Even if if doesn't make it into current release, it is worth looking at a future one. I think that progression has had side benefits, of us discovering other issues, gotchas and problems in the language specification.

gafter commented:
@AdamSpeight2008 Because this pull request is so large, and has been presented in numerous revisions, I have no idea which parts have been reviewed and which not, and which parts have outstanding issues. I'd prefer to see such changes presented in a series of separate PRs each focused on a particular set of issues.

gafter commented:
@AdamSpeight2008 I do not have any idea why @jasonwilliams200OK was giving you instructions of what to do with this PR.

AdamSpeight2008 commented:
@gafter They were helping me reduce the number of commits ( I think it was in the region of 20) in PR, since I seem to break thing when I do git commands.

AdamSpeight2008 commented:
Quick overview of concern and issues
- [x] Formatting Changes  
*None / minor / end of line whitespace removed.*

- [x] Use of Inlining attributes  
*Removed and not used.*

- [x] Unified methods `ScanTokenFullWidth` and `TryScanToken`
*implemented, Tested, Working. Size reduction of the sourcecode and IL code outputted*

- [x] Implement and use `TryPeek`  
*Completed, iteratively tested (Pass)*

- [x] Issue around usage of contractions in method names.  
*Methods now use the non-contracted forms.* 

- [x] Test Failures do to Locale #1006  
*Doensn't occur now (see #1889)*

- [ ] JIT Inlining

- [ ] Performance

gafter commented:
@AlekseyTs Can you please make an executive decision as to whether we want to move forward with this PR, or instead close it and request that the submitter re-submit smaller more focused PRs for the elements of it?

AdamSpeight2008 commented:
Again?!  True it's a big PR.

The `Scanner(Unification)` change affect 2/3 methods, but as the `TryParse` changes developed some of those changes got incorporated into the `Scanner(Unification)` methods. Submitting (yet another set of) PRs I don't think would be "better". If the number of comments on this PR (and previous) are anything to go by,  have helped refine / improve the changes. It has stabilised and there hasn't been a changes in a while, if there anything that may need altering or change could be fixed in smaller subsequent Pull Requests.

gafter commented:
@AdamSpeight2008 One reason for the decreasing number of review comments and changes is that some reviewers have grown tired of re-reviewing thousands of lines of code multiple times as you revise the PR. This code has not been reviewed since any of the last 10 revisions (perhaps more), and I am not motivated to review it again because it is so large and keeps changing out from under me.

In any case, my question was for @AlekseyTs.

AdamSpeight2008 commented:
@gafter  Oh don't treat me like a doormat, to be walked all over.

It's not like the `master` branch has not changed under me as well.
- I've had all of the variable names change.  (after the PR was submitted )
Which invalidated PR merge would now be a build failure, prior it wasn't.
- Rebase to current state of `master`
- Squash the commits 
Both of which came for "Team" members. "Team" member wanted changes / improvements which I've complied with. *Hence additional commit to this PR*
- Been instructed to not to submit another PR that is essence this / these.
- Recent commits are only to rebase to 'master`

See it from me side.
If I do the follow imaginaryl PRs
```
  1. Scanner Methods (Unified)
  2. TryParse Methods
```
   Merge conflict as the 2nd contains the older methods. Thus I have rebase if
or
```
  1. TryParse Method
  2. Scanner Methods (Unified)
```
  Merge conflict as the unified method doesn't contain the use of TryParse. Thus require addition commit / rebase.

Either way I'm screwed i terms merge conflicts

Alternative this
- I submit yet another PR, lets say Scanner Methods (Unified).
- Wait for the Yah / Nah.
- Wait for the merge into master.
- Fork with this new master as a base,
- Redo all the TryParse implememtations. ( possible missing some / making errors )
- Submit yet another PR.
- Wait for the Yah / Nah.

Wouldn't it be simpler for everyone to have one PR (this one)?

jaredpar commented:
@AdamSpeight2008 

The issue with this PR is that it's a collection of changes rather than a single change to the scanner:

- Formatting changes: white space removal, line break changes, etc ... 
- Refactoring: new helper methods: `Peep`, `IsnotFrom`, etc ... , rewriting core methods like `ScanTokenFullWidth`
- Code hygene changes: switching to interpolated strings

Combining all of these changes into a single PR, and a large one at that, makes it significantly costlier for the Roslyn team to review and verify.  The ask from @gafter, which I completely agree with, is for the PR to be broken into smaller focused PRs.  

In its current form I don't see any way we could merge this PR into the source tree: 

- The scanner is a **very** perf sensitive area for the compiler and we don't treat refactoring changes to it lightly. Verifying the change involves careful monitoring of our perf suites plus an investigation to any deviations we see.  This is why we are pushing to separate out the refactorings from all of the other changes such as formating, name changes, etc ... It simplifies the investigation process to have the change as minimal as possible.  
- I know you disagree with our formatting and style decisions but they are not going to change and we will not be accepting PRs that contain arbitrary reformattings of our code.  I've mentioned this several times in the past.  
- The size of the change and mixed refactorings make it harder to follow.  It's difficult to tell if parts of the change are bug fixes or changes that respond to refactorings you made.  

There is a general question as to whether we would take the refactoring of the Scanner even if it was separated out from the rest of the change.  Some of the pattern changes, like `Peep`, look good at least to me.  Other changes are of questionable value and have a high cost for the team in terms of verifying the changes.  Breaking the refactoring into smaller, easier to review, changes would significantly increase the likely hood we would accept it into our source tree. 

I do want to thank you for pushing on this change.  We are interested in adding value to our code bases and cleaning up parts of our code base which have gone stale.  I appreciate that you've put a good deal of effort into this PR and are trying to add value to our code base.  But in its current form it's not possible for us to accept it for the reasons listed above.  Given all of this I'm closing this PR.
